### ISSUE TITLE
Storage becomes the coupled-sector matrices; recoupling becomes a matmul

### ISSUE THREAD

**Issue by Ryo-wtnb11** [Ryo-wtnb11/TeNeT-py#325]
Rewritten after reading TensorKit.jl 0.17.1, symmray and frostspin at source level. This
issue now fixes the design rather than proposing a direction; the earlier body and its
first correction are kept in the comments.

## 1. What the three libraries actually do

| | TensorKit.jl | frostspin | symmray (sparse / flat) |
|---|---|---|---|
| symmetry | non-abelian, general | non-abelian (SU(2), O(2)) | abelian + fermionic parity only |
| storage | **one flat `DenseVector`**; coupled-sector blocks concatenated into it | **tuple of dense matrices**, one per coupled irrep, sorted | dict `sector -> block` / one stacked `(num_blocks, *shape)` array |
| a block is | `StridedView(data, sz, str, offset)` from a precomputed table | a row/column chunk of its irrep matrix | a dict value / a row of the stack |
| tree vs degeneracy | tree index selects the offset; degeneracy are the view's axes. **Tree slow, degeneracy fast** | row index is `i_struct * ext + i_ext` inside a per-configuration chunk. **Structural slow, external fast** | no fusion trees exist |
| recoupling | pack the trees of a group into a scratch matrix, **one `mul!` against `Uᵀ`**, scatter | strided view + one transpose, **one `unitary_rows @ swapped_old_sym_mat`**, scatter | not applicable |
| contraction | **one `mul!` per coupled sector** over a merge-join of sorted block iterators | **one `@` per shared irrep** over a merge of sorted irrep lists | fuse -> blockwise/batched matmul -> unfuse, or `direct` with a lexsort |
| caches | `sectorstructure` keyed on **sectors only, degeneracy-independent**; `degeneracystructure` per space; `treebraider`/`treetransposer` per (spaces, permutation); `fsbraid` per tree block. All LRU, default 10⁴ | `_structural_data_dic` keyed on **irrep labels without degeneracies**; `_unitary_dic` per elementary block. Unbounded, process-lifetime | fuse-info LRU with a size cap **and a "too many sectors, do not cache" bail-out**, plus hit/miss counters |
| stated cost model | **memory movement, not flops** (`_contract_memcost` counts bytes permuted and takes the gemm as fixed). Docs: fusion-tree combinatorics is amortized to ~zero by caching, so runtime is BLAS plus data movement | contraction order chosen assuming external multiplicity dominates irrep dimension — the PEPS assumption | fuse bookkeeping is the sparse hot spot; `lexsort` is the flat one (`# XXX: this is a bottleneck`) |

Three points where they agree, and agreement across three independent implementations is
the strongest evidence available here:

1. **A block is a view into a larger contiguous object, never an array of its own.** No
   library gathers blocks to operate on them, because none of them has to.
2. **The fusion-tree index is the slow axis and the degeneracy the fast axis.**
   TensorKit's `offset₁ += d₁` over the splitting structure and frostspin's
   `(struct, ext, struct, ext)` reshape are the same statement.
3. **Recoupling is one dense `gemm` per tree group against a cached change-of-basis
   matrix.** Neither touches an F-symbol or a Clebsch-Gordan coefficient at run time.

## 2. What tenet cannot take, and why

This is the part that is ours, not theirs: tenet reaches NumPy, JAX and PyTorch through
autoray, and two of the three designs above depend on capabilities that set does not have.

- **TensorKit's flat vector plus a stride table is out.** It works because every subblock
  is `StridedView(data, sizes, strides, offset)` — an arbitrary strided view. JAX has no
  `as_strided` (verified), and a traced program cannot construct one. NumPy and PyTorch
  can; JAX cannot, so it is not the storage.
- **frostspin's layout is in.** Its blocks are recovered by a **2-D slice plus a reshape**
  of a per-sector matrix, which JAX does support (verified), and which #326 already
  established is a view on NumPy because the reshape only ever *splits* axes.
- **numba is out** (no new dependencies), and the frostspin review says it is not needed
  for this: of its ~22 `njit` functions only three do irreducible per-coefficient work,
  all three are in the *abelian/O(2)* paths that never received the layout fix, and two
  more are dead code. The Lie-group path's numba is Python-overhead avoidance around BLAS.
- **symmray's `flat` backend is out.** It requires uniform block shapes *and* a finite
  cyclic symmetry with charges `0..N-1`: its own docs say U(1) is unsupported. It is a
  narrow specialization for `jax.jit`, not a general layout.
- **In-place assembly is already conditional and stays so.** symmray splits the same way
  (`zeros`+insert on NumPy, `concatenate` elsewhere, "more compatible with e.g. autodiff
  since requires no inplace updates"), and tenet's `_MUTABLE` split predates this issue.

### 3.1 Storage

A `SymmetricTensor` holds **one dense matrix per coupled sector**, in coupled-sector
order — frostspin's layout, which is the portable one.

```
data: tuple[Array, ...]        # one matrix per coupled sector, sector-sorted
```

`blocks` stays public and keeps its meaning — `blocks[i]` belongs to `block_order[i]` —
but becomes a **derived property** returning views, which is what `from_matrices` already
computes and what #326 made cheap. Design invariant #8 is amended, not dropped: the
storage contract becomes *"`data` is a tuple of per-coupled-sector arrays; `blocks` is the
public view of it, and both are array-free at the structure level"*. The immutability and
hashability of structures is untouched.

For JAX this is strictly better: the pytree flattens to a handful of large leaves instead
of tens of thousands of small ones, and the per-block slices whose VJPs produced the 3,538
single-primitive XLA compilations measured in #322 disappear.

### 3.2 Index order inside a sector matrix

Rows are ordered by **(irrep configuration, tree within configuration, degeneracy)**:

```
row index = offset(configuration) + i_tree * extent(configuration) + i_degeneracy
```

with `extent(configuration) = Π_k m_k` over that side's legs. Columns likewise. An *irrep
configuration* is one choice of uncoupled irrep per leg on that side; within one
configuration every block has the same external shape **by construction**, and the
remaining freedom is which fusion tree — the structural multiplicity.

This is the ordering both TensorKit and frostspin use, and it subsumes #326's
sort-by-extent: blocks of one configuration share an extent automatically, so the
rectangles that #326 walks are exactly the configuration stripes, and they stay
contiguous. It is strictly finer than #326's, which grouped merely-equal extents that may
carry different shapes.

### 3.3 Metadata and its cache keys

Two tables, both pure functions of the structure, both cached, split exactly as TensorKit
splits `sectorstructure` from `degeneracystructure`:

- **`tree_structure`** — the configurations, the trees in each, the coupled sectors. Keyed
  on the **sector content only, with degeneracies excluded**. Both TensorKit
  (`Hashed(W, sectorhash, sectorequal)`) and frostspin (`r[1:].ravel()`, labels only) do
  this deliberately, and our own measurement agrees: the structural count is 1,967 at
  SU(2) rank 6 whether the degeneracies are `2` or `j+1`. One entry then serves every bond
  dimension.
- **`layout`** — offsets and extents. Degeneracy-dependent, keyed on the full structure.

### 3.4 Recoupling

A repartition or permutation becomes, per (source configuration, destination
configuration):

1. one transpose of the source stripe viewed as `(tree, *external axes)`, which moves the
   structural axis to the front **and** reorders the external axes into the new leg order
   in the same call — frostspin's `data_perm` does both jobs at once;
2. one `U @ M`, with `U` of shape `(n_dst_trees, n_src_trees)` **cached on the
   degeneracy-independent key**;
3. one write of the result into the destination stripe.

`terms` — the `(source block, destination block, coefficient)` list — is the sparse COO
form of exactly these `U`. It stops being walked and becomes the thing `U` is built from,
once per distinct plan.

### 3.5 Operations

- **`tensordot`**: one matmul per coupled sector over a merge-join, and nothing else. The
  `to_matrices`/`from_matrices` round trip disappears because the matrix *is* the storage.
  A partition that already matches costs nothing — TensorKit's `has_shared_permute`, and
  tenet's #327 identity skip is the same observation reached from the other end.
- **`svd`/`qr`/`eigh`**: unchanged in meaning, one call per sector, already the shape they
  have.
- **`to_dense`/`from_dense`, `serialize`**: unchanged in contract. `serialize` writes
  blocks today and can keep doing so, or write `data`; that is a format decision, not a
  design one.

### 3.6 Cache policy

Take symmray's, not the other two's. tenet's plan caches are unbounded `@cache`; frostspin's
are unbounded and process-lifetime; TensorKit's are LRU at 10⁴. symmray additionally
**declines to cache when there are too many sectors** and instruments hit and miss counts.
A structure-keyed cache that grows without bound is a leak in a long optimization run, and
we have no measurement of hit rates. Bound them, and count.

## 4. Staging

1. **`tree_structure` and the configuration ordering**, with `blocks` still materialized.
   No behaviour change; the ordering becomes the configuration order and #326's rectangles
   become configuration stripes. Measurable on its own.
2. **Storage becomes the sector matrices**, `blocks` becomes a view property. This is the
   invariant-#8 amendment and the pytree change. Everything downstream keeps working
   because `blocks` still answers.
3. **Recoupling becomes `U @ M`.** Plans build and cache `U` instead of `terms`; `terms`
   remains as the sparse form the builder emits.
4. Delete the fuse/unfuse round trip, which by then has no callers.

Each stage is separately landable and separately measurable. Stage 2 is the decision this
issue asks for; stages 1 and 3 are consequences.

## 5. Acceptance

- Bit-identical results on every provider, both degeneracy patterns, all three backends.
  The existing round-trip and dense-oracle tests are the criterion and are not to be
  weakened.
- A test that the recoupling issues one matmul per (configuration pair), not one call per
  block — the `count_backend_calls` shape used in #319 and #326.
- No regression on the JAX-traced CTMRG gradients (`test_ctmrg_ipeps.py`,
  `test_ctmrg.py::test_main_runs_both_halves`), which caught #319's.
- The degeneracy-independence of `tree_structure` asserted directly: the same key for
  degeneracies `2` and `j+1`.

## 6. Where this leaves the numbers

After #319, #322, #323, #326 and #327, `tensordot` on SU(2) rank 6 with 9,145 blocks is
**217 ms against TensorKit's 243.7 ms**, and the raw `gemm`s alone are 77.6 ms. The
remaining ~140 ms is the fuse/unfuse round trip and the gather that feeds every batched
operation — which is what this design removes. Small tensors will stay far behind
regardless: at rank 4 TensorKit is 0.005 ms against our 0.32 ms, and that is Python's
per-call floor, not a layout problem.

**Comment by Ryo-wtnb11** [comment]
Correcting this issue's premise with measurements taken after #326 and #327.

**The term walk is no longer the cost, and the reason is not the one stated above.**

Two measurements changed the picture.

**1. An unbent contraction's restore is the identity.** 613,468 terms, every one
`(i, i, 1)`, identity permutation, unchanged structure — walked to hand back the tensor
they were read from, 0.4 s of a 0.47 s `tensordot`. That is not recoupling and it is not
what this issue is about; #327 skips it.

**2. Where a real recoupling remains, `batch_plan` already collapses it.** On a
contraction that bends one leg: 34,673 terms, of which **33,431 fall into a single
batched group and 1,242 stay on the loop**. The Python-per-term problem this issue opens
with is already solved by #319.

What actually dominates the bent case now (profile of 20 calls, 15.8 ms each):

```
0.229  apply_plan                          (72% cumulative)
0.052    numpy stack
0.019    scaled                            (25,840 calls)
0.037  SymmetricTensor.__post_init__       (2 per tensordot, 9k block shapes each)
       repartition.py genexprs             182,920 and 158,080 calls
```

`batch_plan`'s execution has to **`stack` the source blocks of a bucket** — 9,146 of them
per call — because they are separate array objects. The stack is a copy of the whole
bucket, and building its argument is a 9,146-element Python generator.

**That is the argument for the factorization, restated correctly.** If the blocks sharing
an irrep configuration were already contiguous in one array, the bucket would be a *view*
of storage and the stack would be a reshape. The win is not "the term loop becomes a
gemm" — the term loop is gone — it is "the gather that feeds every batched operation
disappears".

That also sharpens what the factorization has to be. Grouping by *shape*, which is what
`batch_plan` does, is not enough to make storage contiguous, because the blocks of one
shape are scattered. The grouping that makes storage contiguous is by **irrep
configuration**: the uncoupled irrep on each leg. Within one configuration every block
has the same external shape by construction, and the remaining freedom is which fusion
tree — the structural multiplicity. That is exactly frostspin's
`structural x external` split, and it is a property of the structure, so it can be the
storage order.

Secondary, and cheap: `SymmetricTensor.__post_init__` validates every block's shape and
is 12% of the bent case (3% at rank 6 unbent). Internal constructions that a plan just
built have already established their shapes. Worth a separate look; the trust boundary is
deliberate, so it is a design question, not an optimization.

**Comment by Ryo-wtnb11** [comment]
Implemented by #331, all three stages.

Verified against `origin/main` `0e91c10`, three rounds each, and independently of the implementation report:

| | main | after |
|---|---|---|
| belief propagation, warm ms/iteration | 165.2 / 162.9 / 163.0 | **139.0 / 138.1 / 138.4** |
| `test_ctmrg_ipeps.py` wall (JAX) | 56.7 / 57.9 s | **50.8 / 54.1 s** |
| `tests/backends/test_torch.py` | 148 passed, 2 failed | **150 passed** |
| traced jaxpr, su2 cold gradient | 42,867 equations, 1,550 `pad` | 44,054, 1,767 |
| `dot_general` (the arithmetic) | 337 | **337** |
| tensordot uniform rank 5 / 6 | 3.695 / 217.99 ms | **0.379 / 89.5 ms** |

Installed in the driving project: **134 ms/iteration**, from ~200 where this started.

Two things the issue got wrong, both corrected in its comments rather than quietly:

- **Stage 2 alone made BP 4-5% slower.** The issue attributed 83% of BP to the round trip stage 2 deletes; it moves the round trip instead, because BP's hot path is handed blocks and produces blocks. The removal is stage 3.
- **The first stage-3 attempt made the JAX path 4-5x slower** and took a CI runner down. With matrices as storage, *what an operation reads becomes its cost on a traced backend* — a block is a view on NumPy and a `slice` with a `pad` behind it on JAX. symmray answers this with two backends; the finding here is that the split is needed not at the layout level but at the level of which operations are matrix-native. Once they are, one layout serves both.

Two structural tests enforce that going forward — counts of primitives in a traced jaxpr, not durations: a matrix-native operation cuts **zero** blocks into the graph, and an operation that reads blocks does not also gather them back.

Left open deliberately, as follow-ups rather than parts of this:

- `repartition` measured alone is 1.2-1.5x above main, because `main` never did the work — it left unit-coefficient terms as transposed views and deferred the copy. The copy is real either way in a pipeline, which is why the pipeline is 15% faster. Attacking it means one strided write per configuration rather than per block.
- The remaining +2.8% of traced equations is the plan applier: on an immutable backend `lower_plan` cannot write in place, so `apply_plan` is handed blocks. A recoupling that is matrix-to-matrix on immutable backends too is the next step.
- `_restore_plan` and `adjoint_plan` may admit the pattern split #330 found, which would remove their cache growth rather than bounding it. Worth re-checking now that the storage changed.
